lib/wallet.js

```diff
@@ -0,0 +1,191 @@
+        ]);
+
+        const currencies = new Map();
+        currencies.set('0X0000000000000000000000000000000000000000', {
+            symbol: 'ETH',
+            decimals: 18
+        });
+        for (let t of supportedTokens) {
+            currencies.set(t.currency.toUpperCase(), {
+                symbol: t.symbol,
+                decimals: t.decimals
+            });
+        }
+
+        const striimBalance = {};
+        for (let b of striimBalances) {
+            const currency = currencies.get(b.currency.toUpperCase());
+            striimBalance[currency.symbol] =
+                ethers.utils.formatUnits(b.amount, currency.decimals);
+        }
+
+        return striimBalance;
+    }
+
+    /**
+     * Deposits ETH from the on-chain balance of the wallet to striim.
+     * @see https://docs.ethers.io/ethers.js/html/api-providers.html#transaction-receipts
+     * @param {(number|string)} amountEth The amount of ETH to deposit.
+     * @param [options]
+     * @returns {Promise<Object>} A promise that resolves into a transaction receipt.
+     */
+    async depositEth(amountEth, options) {
+        const amountWei = ethers.utils.parseEther(amountEth.toString());
+        const clientFund = _clientFund.get(this);
+        dbg('Sending ' + amountWei + ' wei to ' + clientFund.address + ' from ' + this.address);
+        const tx = await this.send(clientFund.address, amountWei, options);
+        return getTransactionReceipt.call(this, tx.hash);
+    }
+
+    /**
+     * Deposits a token from the on-chain balance of the wallet to striim.
+     * @see https://docs.ethers.io/ethers.js/html/api-providers.html#transaction-receipts
+     * @param {(number|string)} amount The amount of currency to deposit.
+     * @param {string} symbol The currency symbol
+     * @param [options]
+     * @returns {Promise<Object[]>} A promise that resolves into an array of transaction receipts.
+     */
+    async depositToken(amount, symbol, options) {
+        const tokenInfo = await getTokenInfo.call(this, symbol);
+        const amountBN = ethers.utils.parseUnits(amount.toString(), tokenInfo.decimals);
+        const tokenContract = new Erc20Contract(tokenInfo.currency, this);
+        const clientFund = _clientFund.get(this);
+
+        try {
+            var approvalTx = await tokenContract.approve(clientFund.address, amountBN, options);
+            dbg('ApproveTX: ' + JSON.stringify(approvalTx));
+
+            var approvalTxReceipt = await getTransactionReceipt.call(this, approvalTx.hash);
+            dbg('ApproveTX (receipt): ' + JSON.stringify(approvalTxReceipt));
+        }
+        catch (err) {
+            dbg(err);
+            throw new Error('Failed to approve ERC20 token payment in time!')
+        }
+
+        if (approvalTxReceipt.status === 0)
+            throw new Error('Pre-approval transaction failed!');
+
+        try {
+            var depositTx = await clientFund.depositTokens(tokenContract.address, amountBN, options);
```

Comment:
var? Is it to avoid using block scoping of the variables? In that case, wouldn't it be more intuitive to declare the variables in advance, outside the block?

Reply:
Yes it was intentional to "hoist" the variable. It is a bit annoying with the new try..catch semantics of promises that you have to pre-declare variables outside the test scope to keep using them in some cases (i.e. I can move line 105+106 inside the try block since it (rightly so) gives a warning/error about using exceptions for flow control. Good ol' var still has it's uses :P

Reply:
Good catch noticing it in a PR of this size :D